File: P2.py
# TC : O(2n) = O(n)
# SC : O(1)
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        find_size = 0
        dummy = head
        while dummy:
            find_size += 1
            dummy = dummy.next
        # print(find_size)
        i = 1
        dummy2 = head
        if find_size-n==0:
            head = head.next
            return head
        while i != find_size-n:
            # print(dummy2.val)
            dummy2 = dummy2.next
            i += 1
        dummy2.next = dummy2.next.next
        return head

File: p2.java
// Time Complexity : O(n)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :Yes
// Any problem you faced while coding this :No


// Your code here along with comments explaining your approach
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        //creating a dummy head
        ListNode dummy = new ListNode(-1);
        dummy.next = head;

        //Creating two pinters slow and fast which start from dummy head and pointing actual head to dummy head as well
        ListNode slow = dummy;
        ListNode fast = dummy;
        head = dummy;

//Fast pointer would go that many steps ahead as n. Using count to do that
        int count = 0;
        while(count != n){
            fast = fast.next;
            count++;
        }
//Iterate fast and slow pointer till fast reached the last node
        while(fast.next != null){
            fast = fast.next;
            slow = slow.next;
        }
//Removing the node next to slow pointer
        ListNode temp = slow.next;
        slow.next = slow.next.next;
        temp.next = null;
//Pointing head again to the original head and returning head
        head = head.next;

        return head;
    }
}